src/OrchardCore.Modules/OrchardCore.ContentPreview/Handlers/PreviewPartHandler.cs

```diff
@@ -41,18 +43,37 @@ public override async Task GetContentItemAspectAsync(ContentItemAspectContext co
 
             if (!String.IsNullOrEmpty(pattern))
             {
-                await context.ForAsync<PreviewAspect>(async previewAspect =>
+                string previewUrl = null;
+
+                var model = new PreviewPartViewModel()
                 {
-                    var model = new PreviewPartViewModel()
-                    {
-                        PreviewPart = part,
-                        ContentItem = part.ContentItem
-                    };
+                    PreviewPart = part,
+                    ContentItem = part.ContentItem
+                };
+
+                previewUrl = await _liquidTemplateManager.RenderAsync(pattern, NullEncoder.Default, model,
+                    scope => scope.SetValue("ContentItem", model.ContentItem));
 
-                    previewAspect.PreviewUrl = await _liquidTemplateManager.RenderAsync(pattern, NullEncoder.Default, model,
-                        scope => scope.SetValue("ContentItem", model.ContentItem));
+                previewUrl = previewUrl.Replace("\r", String.Empty).Replace("\n", String.Empty);
 
-                    previewAspect.PreviewUrl = previewAspect.PreviewUrl.Replace("\r", String.Empty).Replace("\n", String.Empty);
+                await context.ForAsync<PreviewAspect>(previewAspect =>
+                 {
+                     previewAspect.PreviewUrl = previewUrl;
+                     return Task.CompletedTask;
+                 });
+                await context.ForAsync<ContentItemMetadata>(metadata =>
+                {
+                    if (metadata.DisplayRouteValues == null)
+                    {
+                        metadata.DisplayRouteValues = new RouteValueDictionary {
```

Comment:
This should probably call `contentManager.PopulateAspect<PreviewAspect>` inside this func, rather than the change you made to generate the preview url 

Aspects are intended to be extensible so another handler can jump in and replace the aspect values if it wants to later in the pipeline.

You'll need to resolve the contentmanager lazily from a service provider.

Otherwise seems like a good idea

Reply:
@deanmarcussen yes, here, the code is using `GetContentItemAspectAsync` of `PreviewPartHandler`.  Here PreviewPartHandler is also adding "PreviewUrl" in route metadata.

`ContentLinkTagHelper` is using route metadata  for links and calling aspect as below.

`metadata = await _contentManager.PopulateAspectAsync<ContentItemMetadata>(DisplayFor);`

Reply:
Yes, but you can't run this code outside the func, or it will run on many requests, rather than just when the aspect is called.

It would be ok if the code was in a private method that was shared between the two aspects, and called inside the func.

This shouldn't add to the route values either, as otherwise it will actually become a route value for other uses of `ContentItemMetadata`.

It would need to be a seperate property on `ContentItemMetadata`, maybe `DisplayUrl` which could then be checked seperate of the route values.

Reply:
Or it shouldn't be a property on `ContentItemMetadata` at all, and just look for a `PreviewAspect` in the `display-for` helper

Reply:
> It would be ok if the code was in a private method that was shared between the two aspects, and called inside the func.

Understood, will create private method.


> This shouldn't add to the route values either, as otherwise it will actually become a route value for other uses of `ContentItemMetadata`.

I opted for `ContentItemMetadata` as `ContentLinkTagHelper` is using it and I can't use `PreviewAspect` as its not good idea to add reference of `ContentPreview` module in  `OrchardCore.Contents.TagHelpers` for `PreviewAspect` type

Reply:
@deanmarcussen If I move `PreviewAspect` class from `ContentPreview` module to `ContentManagement.Abstractions`  then I can use `PreviewAspect` in `display-for` tag helper.